staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -179,7 +184,15 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 					panicCh <- r
 				}
 			}()
-			list, err = r.listerWatcher.List(options)
+			// Attempt to gather list in chunks, if supported by listerWatcher, if not, the first
+			// list request will return the full response.
+			pager := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
+				return r.listerWatcher.List(opts)
+			}))
+			if r.WatchListPageSize != 0 {
+				pager.PageSize = r.WatchListPageSize
+			}
+			list, err = pager.List(context.Background(), options)
```

Comment:
What I'm a bit afraid of now, is that now we don't have any fallback if this paginated list won't work.
Maybe it's fine given the early phase of release cycle, but this is pretty crucial part (if this won't initialize, watch pretty much won't work at all) so I think we need to be careful.

Reply:
Ah, I should have pointed this out more explicitly: `pager.List` automatically falls back to full list (see pager.List impl), which is what we do today. So the behavior will be that first we attempt a paginated request and if that fails we fall back to full list. If both fail the cacher retries using the same flow it has today.

Reply:
I looked into code and it fallbacks only if it fails due to "expiration" error.
I think I can live with that, but at least let's add a comment about the semantics both here and an the watchcache code.

Reply:
Sounds good, adding.

Reply:
Comments added. Thanks!

As you might already know, in addition to the fallback, there is a `wait.Until` `cacher.startCaching` that will loop that retries the list operation, so if the failure is transient, the cacher will attempt to recover.

Reply:
> As you might already know, in addition to the fallback, there is a wait.Until cacher.startCaching that will loop that retries the list operation, so if the failure is transient, the cacher will attempt to recover.

Yeah - I know. But if regular list would work and paginated one no, we would never recover.

Reply:
I am not aware of any case that could cause that that isn't "etcd has a horrible bug".  I'm not saying that should be a blocker, but we'd break in other places too, like all informers.

Reply:
So to be clear - I'm also not expecting any issues with this - it just the "being on the safe size" thing.